### PR TITLE
Feature/post processing

### PR DESCRIPTION
**Pull request by troehling**
Implementation of new processing pipeline:
- Using the same code for online and post processing
- better performances due to better sizing of the packages
- test case for post processing based on the results of previously used RawConverter
- Using git lfs for storage of large files

### CODE REVIEW

**Review by BenMoon — CHANGES_REQUESTED** [review]
Overall, this looks quite good.
In general, maybe we should employ a linter like `flake8` or such.
Pls check the header in each file.

File: pymepix/main.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        pymepix[0].loadConfig(args.spx)
+
+    # Switch to TOF mode if set
+    if args.decode and args.tof:
+        pymepix[0].acquisition.enableEvents = True
+
+    # Set the bias voltage
+    pymepix.biasVoltage = args.bias
+
+    ext = "raw"
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/main.py:61]: Done. File output needs to be implemented for pymepix! See #23.

File: pymepix/main.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        logging.info("Decoding data enabled")
+        if args.tof:
+            logging.info("Tof calculation enabled")
+            ext = "tof"
+        else:
+            ext = "toa"
+    else:
+        logging.info("No decoding selected")
+
+    output_file = open_output_file(args.output, ext)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/main.py:72]: See #23 regarding output files.

File: pymepix/main.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        logging.debug("Datatype: {} Data:{}".format(data_type, data))
+        if data_type is MessageType.RawData:
+            if not args.decode:
+                store_raw(output_file, data)
+        elif data_type is MessageType.PixelData:
+            if args.decode and not args.tof:
+                store_toa(output_file, data)
+        elif data_type is MessageType.PixelData:
+            if args.decode and args.tof:
+                store_tof(output_file, data)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/main.py:104]: See #24: The functions are not removed because they are still used in pymepix-viewer.

File: pymepix/post_processing.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+class ProgressBar(tqdm):
+
+    def update_to(self, progress):
+        return self.update(progress - self.n)
+
+
+def run_post_processing(input_file_name, output_file, number_processes, timewalk_file, cent_timewalk_file):
+    with ProgressBar(total=1.0, dynamic_ncols=True) as progress_bar:
+        file_sampler = RawFileSampler(input_file_name, output_file, number_processes, timewalk_file, cent_timewalk_file, progress_bar.update_to)
+        file_sampler.run()
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/post_processing.py:14]: Done.

File: pymepix/processing/acquisition.py [hunk 5]
```diff
@@ -36,88 +38,18 @@ class PixelPipeline(AcquisitionPipeline):
     def __init__(self, data_queue, address, longtime, use_event=False, name="Pixel"):
         AcquisitionPipeline.__init__(self, name, data_queue)
         self.info("Initializing Pixel pipeline")
-        self._use_events = use_event
-        self._event_window = (0, 10000)
+        self.packet_processor = PacketProcessor(handle_events=use_event, event_window=(0, 1E-3))
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/acquisition.py:41]: Done.

File: requirements.txt [hunk 6]
```diff
@@ -4,3 +4,4 @@ pyzmq
 h5py
```
[the remark this answers is not in the document]
Comment (troehling) [reply requirements.txt:4]: Added a comment stating that it is only used for GitHub actions. Created an issue to use installation through setup.py in the github actions (#22)

File: tests/test_centroid_calculator.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    y = np.concatenate(([0, 0, 1, 0, -1, 0], np.array([1, 1, 2, 1, 0, 1]) + 5))
+    tof = np.array([0, 0, 0, 0, 0, 0] + [0, 0, 0, 0, 0, 0])
+    tot = np.array([1, 1, 1, 1, 1, 1] + [1, 1, 1, 1, 1, 1])
+
+    expected_result = [1, 2], [0, 6], [0, 6], [0, 0], [1, 1], [1, 1], [6, 6]
+    assertCentroidsEqual(expected_result, centroid_calculator.process((shot, x, y, tof, tot)))
+
+def assertCentroidsEqual(expected, actual):
+    for i in range(len(expected)):
+        np.testing.assert_array_equal(expected[i], actual[i])
```
[the remark this answers is not in the document]
Comment (troehling) [reply tests/test_centroid_calculator.py:177]: Done.

File: tests/test_centroid_calculator.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (troehling) [reply tests/test_centroid_calculator.py:1]: Done.

File: tests/test_post_processing.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (troehling) [reply tests/test_post_processing.py:1]: Done.

File: pymepix/processing/acquisition.py [hunk 10]
```diff
@@ -133,126 +65,25 @@ def __init__(self, data_queue, address, longtime):
...
         )
         self.info("Initializing Centroid pipeline")
-        self._skip_centroid = 1
-        self._tot_threshold = 0
-        self._samples = 5
-        self._epsilon = 2.0
+        self.centroid_calculator=CentroidCalculator()
 
-        self.addStage(4, Centroiding, num_processes=25)
+        self.addStage(4, PipelineCentroidCalculator, num_processes=6)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/acquisition.py:70]: This works only in pymepix-viewer currently. The property numBlobProcesses can be changed. Before it is changed the pipeline is stopped (in the viewer) and started again afterward. Otherwise its not possible to change the number of Processes. In this case the 6 is only the initial value can not be changed from the call of the constructor. This is also not intended by design of the pipeline as the instance is created in some kind of hidden way (Not my implementation that how it already was before).

File: pymepix/processing/acquisition.py [hunk 11]
```diff
@@ -133,126 +65,25 @@ def __init__(self, data_queue, address, longtime):
             self, data_queue, address, longtime, use_event=True, name="Centroid"
         )
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/acquisition.py:66]: Added explanation for this one but I do not know about the other ones more precisely. As far as I know it is related to the the inter process communication.

File: pymepix/processing/baseacquisition.py [hunk 12]
```diff
@@ -324,7 +324,7 @@ def main():
     test_value = Value("I", 0)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/baseacquisition.py:324]: I have no idea about this, but as it is only in the main part of this file I would leave it this way.

File: pymepix/processing/baseacquisition.py [hunk 13]
```diff
@@ -324,7 +324,7 @@ def main():
...
 
     acqpipline.addStage(0, UdpSampler, ("127.0.0.1", 50000), test_value)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/baseacquisition.py:326]: I have no idea about this, but as it is only in the main part of this file I would leave it this way.

File: pymepix/processing/baseacquisition.py [hunk 14]
```diff
@@ -324,7 +324,7 @@ def main():
...
-    acqpipline.addStage(2, PacketProcessor, num_processes=4)
+    acqpipline.addStage(2, PipelinePacketProcessor, num_processes=4)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/baseacquisition.py:327]: Yes. For changing it while the pipeline is running there is another function, which still requires the pipeline to be stoped and restarted.

File: pymepix/processing/logic/centroid_calculator.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+
+from pymepix.processing.logic.processing_step import ProcessingStep
+
+class CentroidCalculator(ProcessingStep):
+
+    def __init__(self, tot_threshold=0, epsilon=2, min_samples=3, triggers_processed=1, chunk_size_limit=6_500, 
+        cent_timewalk_lut=None, parameter_wrapper_class=ProcessingParameter):
+
+        super().__init__("CentroidCalculator")
+        self._epsilon = parameter_wrapper_class(epsilon)
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:16]: There is another wrapper implementation that is required for multiprocessing and which interferes with the mp.Pool used in this class. Therefore the wrapper is required to make this functionality interchangeable. The setters and getters are implemented only for outside access and to hide the "*.value" within the class itself.

File: pymepix/processing/logic/centroid_calculator.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+import numpy as np
+from pymepix.processing.logic.processing_parameter import ProcessingParameter
+import scipy.ndimage as nd
+from sklearn.cluster import DBSCAN
+
+from pymepix.processing.logic.processing_step import ProcessingStep
+
+class CentroidCalculator(ProcessingStep):
+
+    def __init__(self, tot_threshold=0, epsilon=2, min_samples=3, triggers_processed=1, chunk_size_limit=6_500,
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:12]: Done.

File: pymepix/processing/logic/centroid_calculator.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+            if trigger_chunk_voxel_counter < self._chunk_size_limit:
+                trigger_chunk_voxel_counter += unique_trig_nr_counts[index]
+            else:
+                trigger_chunks.append(unique_trig_nr_index)
+                trigger_chunk_voxel_counter = unique_trig_nr_counts[index]
+
+        return trigger_chunks
+
+    def __centroid_chunks_to_centroids(self, chunks):
+        centroids = [[] for i in range(7)]
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:110]: Because in this case this is the number of dimensions which is determined by the `calculate_centroids_properties`-method. This builds up the stack of columns for the centroids (which is initially empty).

This implementation might be bad (not sure) and I can not remember why np.concatenate(chunks) did not work here but there must have been some reason for this. I should check this again i guess.

File: pymepix/processing/logic/centroid_calculator.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+                y[label_filter],
+                tof[label_filter],
+                tot[label_filter],
+                labels[label_filter],
+            )
+
+        return None
+        
+    def perform_clustering(self, shot, x, y, tof):
+        """ The clustering with DBSCAN, which is performed in this function is dependent on the order of the data in rare cases. Therefore reordering in any means can
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:149]: Done.

File: pymepix/processing/logic/centroid_calculator.py [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+
+    def post_process(self):
+        self._pool.close()
+        self._pool.join()
+        return super().post_process()
+
+    def __getstate__(self):
+        self_dict = self.__dict__.copy()
+        del self_dict['_pool']
+        return self_dict
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:235]: Done.

File: pymepix/processing/logic/centroid_calculator.py [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+            # cluster_tof *= 1e6
+            cluster_tof -= self._cent_timewalk_lut[np.int(cluster_totMax // 25) - 1] * 1e3
+            # TODO: should totAvg not also be timewalk corrected?!
+            # cluster_tof *= 1e-6
+
+        return cluster_shot, cluster_x, cluster_y, cluster_tof, cluster_totAvg, cluster_totMax, cluster_size
+
+
+class CentroidCalculatorPooled(CentroidCalculator):
+
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:215]: Done.

File: pymepix/processing/logic/packet_processor.py [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+                )
+            ]
+
+            if pixels.size > 0:
+                pixel_data = self.process_pixels(np.int64(pixels), longtime)
+
+                if triggers.size > 0:
+                    self.process_triggers(np.int64(triggers), longtime)
+
+                if self.handle_events:
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/packet_processor.py:86]: I can not really answer this as I did just copy it over from the old code. My assumption was that handle events is related to tof and centroid calculation and all the other things are also required for toa processing. Therefore I assume we can not simply check this at the beginning of the method as this would break toa processing.

File: pymepix/processing/logic/packet_processor.py [hunk 22]
```diff
@@ -0,0 +1,302 @@
+from enum import IntEnum
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/packet_processor.py:1]: Not sure if it would have been better because this is actually a new class that provides the functionality of the previous packet processor with the same name. The only extracted functionality that is not provided is the inter process communication which is performed in the `pipeline_packet_processor` now.

Header has been added now.

File: pymepix/processing/logic/processing_parameter.py [hunk 23]
```diff
@@ -0,0 +1,12 @@
+class ProcessingParameter:
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/processing_parameter.py:1]: Done for all files in `logic`package.

File: pymepix/processing/logic/processing_step.py [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+
+    def pre_process(self):
+        pass
+
+    def post_process(self):
+        pass
+
+    @abstractmethod
+    def process(self, data):
+        pass
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/processing_step.py:18]: Done.

File: pymepix/processing/logic/processing_step.py [hunk 25]
```diff
@@ -0,0 +1,18 @@
+from abc import abstractmethod, ABC
+
+from pymepix.core.log import Logger
+
+class ProcessingStep(Logger, ABC):
+
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/processing_step.py:6]: Added  the docstring and tried to explain the connection among the classes briefly. Not sure whether a diagram would be helpful because I still do not fully understand all t he things happening in the (already) existing pipeline implementation. Therefore the picture would be rather incomplete. For The additions I implemented the brief explanation should satisfy the need for documentation.

File: pymepix/processing/logic/shared_processing_parameter.py [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+        else:
+            raise UnknownParameterTypeException()
+
+    @property
+    def value(self):
+        return self._value.value
+
+    @value.setter
+    def value(self, value):
+        self._value.value = value
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/shared_processing_parameter.py:24]: Done.

File: pymepix/processing/rawfilesampler.py [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+                elif subheader == 0x5:
+                    should_push = self.handle_msb_time(pixdata)
+
+            if should_push:
+                self.push_data()
+
+        if len(self._packet_buffer) > 0:
+            self.push_data()
+        
+        self.post_run()
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/rawfilesampler.py:272]: Done.

File: pymepix/processing/pipeline_packet_processor.py [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+            PipelinePacketProcessor.__name__,
+            input_queue=None,
+            create_output=create_output,
+            num_outputs=num_outputs,
+            shared_output=shared_output,
+        )
+        self.packet_processor = packet_processor
+
+    def init_new_process(self):
+        """create connections and initialize variables in new process"""
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/pipeline_packet_processor.py:57]: Done. Did simply remove the docstring because it is obvious at a single glance what is done in this method. Such a comment might only lead to confusion if the content of the method is changed again in the future and the docstring is "forgotten" and not adapted.

File: pymepix/processing/baseacquisition.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/baseacquisition.py:324]: Done. Renamed to longtime.

File: pymepix/processing/logic/centroid_calculator.py [hunk 29]
```diff
@@ -0,0 +1,235 @@
...
+            return None
+
+    def __skip_triggers(self, shot, x, y, tof, tot):
+        unique_shots = np.unique(shot)
+        selected_shots = unique_shots[::self.triggers_processed]
+        mask = np.isin(shot, selected_shots)
+        return shot[mask], x[mask], y[mask], tof[mask], tot[mask]
+
+    def __divide_into_chunks(self, shot, x, y, tof, tot):
+        # Reordering the voxels can have an impact on the clusterings result. See CentroidCalculator.perform_clustering string doc for further information!
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:80]: Done.

File: pymepix/processing/logic/processing_step.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/processing_step.py:6]: TODO: @troehling, @BenMoon create a diagram to explain this completely #25

File: pymepix/processing/logic/packet_processor.py [hunk 30]
```diff
@@ -0,0 +1,302 @@
...
+    """x=row, y=-column"""
+    Down = 2
+    """x=-column, y = -row """
+    Right = 3
+    """x=-row, y=column"""
+
+
+class PacketProcessor(ProcessingStep):
+    def __init__(self, handle_events=True, event_window=(0.0, 10000.0), position_offset=(0, 0), 
+                orientation=PixelOrientation.Up, start_time=0, timewalk_lut=None, parameter_wrapper_class=ProcessingParameter):
```
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/packet_processor.py:27]: Done.

File: pymepix/processing/logic/centroid_calculator.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (troehling) [reply pymepix/processing/logic/centroid_calculator.py:110]: Very nice! Was able to replace the complete function with np.concatenate(list(chunks), axis=1). Should be easier to understand and faster at the same time. I will also remove the function call and use the function directly instead.
